.github/workflows/ci.yml

```diff
@@ -19,172 +19,318 @@ jobs:
+          auto-update-conda: true
+      - name: Conda environment
+        # `shell: bash -l {0}` - special way to activate modin environment
+        shell: bash -l {0}
+        run: |
+          conda info
+          conda list
       - run: sudo apt update && sudo apt install -y libhdf5-dev
-      - run: pip install -r requirements.txt
-      - run: python -m pytest modin/pandas/test/test_api.py
+      - name: Api tests
+        shell: bash -l {0}
+        run: python -m pytest modin/pandas/test/test_api.py
+
   test-headers:
     runs-on: ubuntu-latest
     name: test-headers
     steps:
-      - uses: actions/checkout@v1
+      - uses: actions/checkout@v2
         with:
           fetch-depth: 1
-      - uses: actions/setup-python@v1
+      - name: Cache conda
+        uses: actions/cache@v1
         with:
-          python-version: "3.7.x"
-          architecture: "x64"
-      - run: pip install -r requirements.txt
-      - run: python -m pytest modin/test/test_headers.py
+          path: ~/conda_pkgs_dir
+          key: ${{ runner.os }}-python-3.8-conda-${{ github.run_id }}-${{ hashFiles('environment.yml') }}
+      - name: Cache pip
+        uses: actions/cache@v1
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-python-3.8-pip-${{ github.run_id }}-${{ hashFiles('environment.yml') }}
+      - uses: goanpeca/setup-miniconda@v1.6.0
+        with:
+          activate-environment: modin
+          environment-file: environment.yml
+          python-version: "3.8"
+          channel-priority: strict
+          use-only-tar-bz2: true # IMPORTANT: This needs to be set for caching to work properly!
+          auto-update-conda: true
+      - name: Conda environment
+        shell: bash -l {0}
+        run: |
+          conda info
+          conda list
+      - name: Headers tests
+        shell: bash -l {0}
+        run: python -m pytest modin/test/test_headers.py
+
   test-internals:
     runs-on: ubuntu-latest
     name: test-internals
     steps:
-      - uses: actions/checkout@v1
+      - uses: actions/checkout@v2
         with:
           fetch-depth: 1
-      - uses: actions/setup-python@v1
+      - name: Cache conda
+        uses: actions/cache@v1
         with:
-          python-version: "3.6.x"
-          architecture: "x64"
-      - run: pip install -r requirements.txt
-      - run: python -m pytest modin/test/test_publisher.py modin/data_management/test/test_dispatcher.py
+          path: ~/conda_pkgs_dir
+          key: ${{ runner.os }}-python-3.6-conda-${{ github.run_id }}-${{ hashFiles('environment.yml') }}
+      - name: Cache pip
+        uses: actions/cache@v1
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-python-3.6-pip-${{ github.run_id }}-${{ hashFiles('environment.yml') }}
+      - uses: goanpeca/setup-miniconda@v1.6.0
+        with:
+          activate-environment: modin
+          environment-file: environment.yml
+          python-version: "3.6"
```

Comment:
is there any reason why three different "pre-major" steps (api, headers, internals) use three different Python versions?

Reply:
This allows us to get a cache for different versions of python (if we put it in a separate job, as in `push.yml`, this will increase the time of checks)

Reply:
this warrants a comment, IMHO